akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonProxy.scala

```diff
@@ -219,16 +234,41 @@ class ClusterSingletonProxy(singletonPathString: String, settings: ClusterSingle
           log.debug("Trying to identify singleton at {}", singletonAddress)
           context.actorSelection(singletonAddress) ! Identify(identifyId)
       }
+    case Terminated(ref) ⇒
+      if (singleton.exists(_ == ref)) {
+        // buffering mode, identification of new will start when old node is removed
+        singleton = None
+      }
 
     // forwarding/stashing logic
     case msg: Any ⇒
       singleton match {
         case Some(s) ⇒
-          log.debug("Forwarding message to current singleton instance {}", msg)
+          log.debug("Forwarding message type [{}] to current singleton instance", msg.getClass.getName)
           s forward msg
         case None ⇒
-          log.debug("No singleton available, stashing message {}", msg)
-          stash()
+          buffer(msg)
       }
   }
+
+  def buffer(msg: Any): Unit =
+    if (settings.bufferSize == 0)
+      log.debug("Singleton not available and buffering is disabled, dropping message [{}]", msg.getClass.getName)
+    else if (buffer.size == settings.bufferSize) {
+      val (m, _) = buffer.removeFirst()
+      log.debug("Singleton not available, buffer is full, dropping first message [{}]", m.getClass.getName)
```

Comment:
This log and the log above is very useful. Would it be possible to have it on info? At least the first time it happens (and log again if it happens again after the buffer was not full)?

If I put the logger on debug, I will probably get a lot of messages all the time?